### PR TITLE
Remove `jl_getch()` to fix race condition in `getpass()`

### LINKED ISSUE #45797: CI: Intermittent hang in libgit2 challenge prompt test.
E.g.: https://buildkite.com/julialang/julia-master/builds/13170#01819131-f9a7-48e4-926f-23e48419c663

We have an rr trace.
```


LibGit2/libgit2                           (9) \|         failed at 2022-06-23T17:20:31.754
--
  | Error During Test at /cache/build/default-amdci4-5/julialang/julia-master/julia-a375fe8755/share/julia/stdlib/v1.9/LibGit2/test/libgit2.jl:2545
  | Got exception outside of a @test
  | Process timed out possibly waiting for a response. Process output found:
  | """
  | Username for 'https://github.com': foo
  | Password for 'https://foo@github.com':
  | signal (15): Terminated
  | in expression starting at none:1
  | unknown function (ip: 0x7000000c)
  | unknown function (ip: 0x7ff02ddd2733)
  | untraced_syscall_base at /workspace/srcdir/rr/src/preload/syscallbuf.c:369 [inlined]
  | sys_read at /workspace/srcdir/rr/src/preload/syscallbuf.c:2436
  | syscall_hook_internal at /workspace/srcdir/rr/src/preload/syscallbuf.c:3387 [inlined]
  | syscall_hook at /workspace/srcdir/rr/src/preload/syscallbuf.c:3494
  | unknown function (ip: 0x7ff02ddcd58f)
  | unknown function (ip: 0x7ff02ddcd5ee)
  | unknown function (ip: 0x7ff02ddcd60a)
  | __read at /lib/x86_64-linux-gnu/libpthread.so.0 (unknown line)
  | ijl_getch at /cache/build/default-amdci4-3/julialang/julia-master/src/sys.c:538
  | _getch at ./util.jl:260 [inlined]
  | getpass at ./util.jl:272
  | _jl_invoke at /cache/build/default-amdci4-3/julialang/julia-master/src/gf.c:2374 [inlined]
  | ijl_apply_generic at /cache/build/default-amdci4-3/julialang/julia-master/src/gf.c:2575
  | getpass at ./util.jl:284
  | authenticate_userpass at /cache/build/default-amdci4-3/julialang/julia-master/usr/share/julia/stdlib/v1.9/LibGit2/src/callbacks.jl:222
  | credentials_callback at /cache/build/default-amdci4-3/julialang/julia-master/usr/share/julia/stdlib/v1.9/LibGit2/src/callbacks.jl:331
  | jfptr_credentials_callback_56476.clone_1 at /cache/build/default-amdci4-5/julialang/julia-master/julia-a375fe8755/lib/julia/sys.so (unknown line)
  | _jl_invoke at /cache/build/default-amdci4-3/julialang/julia-master/src/gf.c:2374 [inlined]
  | ijl_apply_generic at /cache/build/default-amdci4-3/julialang/julia-master/src/gf.c:2575
  | unknown function (ip: 0x7ff00c882288)
  | #credential_loop#7 at /cache/build/default-amdci4-5/julialang/julia-master/julia-a375fe8755/share/julia/stdlib/v1.9/LibGit2/test/libgit2-helpers.jl:31
  | credential_loop##kw at /cache/build/default-amdci4-5/julialang/julia-master/julia-a375fe8755/share/julia/stdlib/v1.9/LibGit2/test/libgit2-helpers.jl:13 [inlined]
  | #credential_loop#8 at /cache/build/default-amdci4-5/julialang/julia-master/julia-a375fe8755/share/julia/stdlib/v1.9/LibGit2/test/libgit2-helpers.jl:68 [inlined]
  | credential_loop at /cache/build/default-amdci4-5/julialang/julia-master/julia-a375fe8755/share/julia/stdlib/v1.9/LibGit2/test/libgit2-helpers.jl:62 [inlined]
  | credential_loop at /cache/build/default-amdci4-5/julialang/julia-master/julia-a375fe8755/share/julia/stdlib/v1.9/LibGit2/test/libgit2-helpers.jl:62


```

### PR DESCRIPTION
**Pull request by staticfloat**
We accidentally introduced a race condition in `getpass()` by having
`jl_getch()` toggle terminal modes for each keystroke.  Not only is this
slower and wasteful, it allows the kernel to receive keystrokes within a
TTY in canonical mode (where it replaces certain characters [0]) and
then reads from the kernel buffer in non-canonical mode.  This results
in us reading a `0x00` when we expected a `0x04` in certain cases on CI,
which breaks some of our tests.

The fix is to switch the TTY into raw mode once, before we ever print
the password prompt, which closes the race condition.  To do this, we
moved more code from C to Julia, and removed the `jl_getch()` export,
instead providing `jl_termios_size()`.

[0] https://github.com/torvalds/linux/blob/e35e5b6f695d241ffb1d223207da58a1fbcdff4b/drivers/tty/n_tty.c#L1318

### CODE REVIEW

**Comment by staticfloat** [comment]
This should fix the intermittent testing failures seen in the following buildkite logs:

- https://buildkite.com/julialang/julia-master/builds/13574#0181d447-3f60-461b-b586-25a38a2476ae/574-1015
- https://buildkite.com/julialang/julia-master/builds/13552#0181d237-9c81-4b3a-b207-da7b6515c951/577-1043
- https://buildkite.com/julialang/julia-master/builds/13548#0181d1af-00f9-45e4-8ffc-904b61885101/577-1451

File: base/util.jl [hunk 1]
```diff
@@ -257,26 +257,84 @@ graphical interface.
 """
 function getpass end
 
-_getch() = UInt8(ccall(:jl_getch, Cint, ()))
+# Note, this helper only works within `with_raw_tty()` on POSIX platforms!
+function _getch()
+    @static if Sys.iswindows()
+        return UInt8(ccall(:_getch, Cint, ()))
+    else
+        return only(read(stdin, 1))
```
Comment (vtjnash) [inline base/util.jl:265]:
```suggestion
        return read(stdin, UInt8)
```

File: base/util.jl [hunk 2]
```diff
@@ -257,26 +257,84 @@ graphical interface.
...
+cfmakeraw(termios) = ccall(:cfmakeraw, Cvoid, (Ptr{Cvoid},), termios)
+
+function with_raw_tty(f::Function, input::TTY)
+    input === stdin || throw(ArgumentError("with_raw_tty only works for stdin"))
+    fd = 0
+
+    # If we're on windows, we do nothing, as we have access to `_getch()` quite easily
+    @static if Sys.iswindows()
+        f()
+        return
```
Comment (vtjnash) [inline base/util.jl:297]:
```suggestion
        return f()
```

Since you have `return try;catch` on the other branch

File: base/util.jl [hunk 3]
```diff
@@ -257,26 +257,84 @@ graphical interface.
...
+    end
+
+    # Get the current terminal mode
+    old_termios = make_termios()
+    tcgetattr(fd, old_termios)
+    try
+        # Set a new, raw, terminal mode
+        new_termios = make_termios()
```
Comment (vtjnash) [inline base/util.jl:304]:
```suggestion
        new_termios = copy(old_termios)
```

File: base/util.jl [hunk 4]
```diff
@@ -257,26 +257,83 @@ graphical interface.
...
+    ret = ccall(:tcsetattr, Cint, (Cint, Cint, Ptr{Cvoid}), Cint(fd), Cint(mode), termios)
+    if ret != 0
+        throw(IOError("tcsetattr failed", ret))
+    end
+end
+cfmakeraw(termios) = ccall(:cfmakeraw, Cvoid, (Ptr{Cvoid},), termios)
+
+function with_raw_tty(f::Function, input::TTY)
+    input === stdin || throw(ArgumentError("with_raw_tty only works for stdin"))
+    fd = 0
```
Comment (vtjnash) [inline base/util.jl:292]: Might as well make this descriptive?
```suggestion
    fd = RawFD(0)
```

File: base/util.jl [hunk 5]
```diff
@@ -257,26 +257,83 @@ graphical interface.
...
+@enum TCSETATTR_FLAGS TCSANOW=0 TCSADRAIN=1 TCSAFLUSH=2
+
+function tcgetattr(fd, termios)
+    ret = ccall(:tcgetattr, Cint, (Cint, Ptr{Cvoid}), Cint(fd), termios)
+    if ret != 0
+        throw(IOError("tcgetattr failed", ret))
+    end
+end
+function tcsetattr(fd, termios, mode::TCSETATTR_FLAGS = TCSADRAIN)
+    ret = ccall(:tcsetattr, Cint, (Cint, Cint, Ptr{Cvoid}), Cint(fd), Cint(mode), termios)
```
Comment (vtjnash) [inline base/util.jl:283]:
```suggestion
function tcsetattr(fd::RawFD, termios, mode::TCSETATTR_FLAGS = TCSADRAIN)
    ret = ccall(:tcsetattr, Cint, (Cint, Cint, Ptr{Cvoid}), fd, Cint(mode), termios)
```

File: base/util.jl [hunk 6]
```diff
@@ -257,26 +257,83 @@ graphical interface.
...
+
+const termios_size = Int(ccall(:jl_termios_size, Cint, ()))
+make_termios() = zeros(UInt8, termios_size)
+
+# These values seem to hold on all OSes we care about:
+# glibc Linux, musl Linux, macOS, FreeBSD
+@enum TCSETATTR_FLAGS TCSANOW=0 TCSADRAIN=1 TCSAFLUSH=2
+
+function tcgetattr(fd, termios)
+    ret = ccall(:tcgetattr, Cint, (Cint, Ptr{Cvoid}), Cint(fd), termios)
```
Comment (vtjnash) [inline base/util.jl:277]:
```suggestion
function tcgetattr(fd::RawFD, termios)
    ret = ccall(:tcgetattr, Cint, (Cint, Ptr{Cvoid}), fd, termios)
```

**Pull request merged**